include/i2c.h

```diff
@@ -110,16 +110,45 @@ union __deprecated dev_config {
  * These are for internal use only, so skip these in
  * public documentation.
  */
+typedef int (*i2c_slave_write_request_t)(void * priv);
+typedef int (*i2c_slave_read_request_t)(void * priv, u8_t *val);
+typedef int (*i2c_slave_write_done_t)(void * priv, u8_t val);
+typedef int (*i2c_slave_read_done_t)(void * priv);
+typedef int (*i2c_slave_stop_t)(void * priv);
+
+struct i2c_slave_api {
+	i2c_slave_write_request_t write_request;
+	i2c_slave_read_request_t read_request;
+	i2c_slave_write_done_t write_done;
+	i2c_slave_read_done_t read_done;
+	i2c_slave_stop_t stop;
+};
+
 typedef int (*i2c_api_configure_t)(struct device *dev,
 				   u32_t dev_config);
 typedef int (*i2c_api_full_io_t)(struct device *dev,
 				 struct i2c_msg *msgs,
 				 u8_t num_msgs,
 				 u16_t addr);
+typedef int (*i2c_api_slave_attach_t)(struct device *dev, u8_t address,
+				      const struct i2c_slave_api *funcs,
+				      void *priv);
+typedef int (*i2c_api_slave_detach_t)(struct device *dev, u8_t address,
+				      void *priv);
 
 struct i2c_driver_api {
 	i2c_api_configure_t configure;
 	i2c_api_full_io_t transfer;
+	i2c_api_slave_attach_t slave_attach;
+	i2c_api_slave_detach_t slave_detach;
+};
+
+typedef int (*i2c_slave_api_attach_t)(struct device *dev);
+typedef int (*i2c_slave_api_detach_t)(struct device *dev);
+
+struct i2c_slave_driver_api {
```

Comment:
I was thinking of moving this in a separate i2c-slave.h header. What were thinking ?

Reply:
There are many other device drivers which relay on I2C API. Sensor drivers would be a good example. Yet, we do not place any sensor related functions in the main I2C API. The drivers which relay on the slave part of I2C API, like eeprom, are no different.  We should keep APIs of different drivers clean and do not mix them.

Also, the next driver using I2C slave interface may choose to attach itself to the I2C bus driver during initialization phase and do not support dedicated attach/detach functions. I don't think we should enforce any specific API design on the device drivers using I2C slave interface. It's best to simply remove these functions from I2C API.

Reply:
Ok, I got your point, you think I should move these attach/detach to eeprom.h header and make them specific to the slave driver.

Yes I was thinking about auto-attach drivers.

@tbursztyka what do you think about that ?

Reply:
I would keep all in i2c.h (slave or master API, it's still i²c APIs anyway)

I don't see how you will register/unregister (these names are nicer than attach/detach actually), if there is no such functions available on the i²c controller side. It's not a slave API here, actually name could be: 

- i2c_api_register_slave
- i2c_api_unregister_slave

that's semantically better.

auto-attach, why not yes.

Question: it's not going to be possible for an i²c controller to serve 2+ i2c slave drivers, here it's one. Would that make sense to actually handle that use case?

If it does, you will have to take a look at how - for instance - I did that in gpio. The callbacks could be in a dedicated, with a sys_node_t in the beginning. 
And the address it would serve as well.

So for the controller it would be a matter of going through that list and find a match on the address, to call the right slave callbacks. What do you think?

Also, it would save the need of passing a void *priv (or context) pointer to the callbacks. Just pass the callback sturct pointer itself, with CONTAINER_OF(), up to the slave driver to get the context back, as it's done in gpio.

Reply:
> I don't see how you will register/unregister (these names are nicer than attach/detach actually), if there is no such functions available on the i²c controller side.

My comments relate strictly to `struct i2c_slave_driver_api` and related functions such as: `i2c_slave_driver_attach`, `i2c_slave_driver_detach`. These belong to the API of the slave driver (which may choose not to implement them or implement differently).

The functions: `i2c_slave_attach`, `i2c_slave_detach` should most certainly stay here.

> actually name could be:
> - i2c_api_register_slave
> - i2c_api_unregister_slave

Zephyr's API is a bit chaotic, but the convention: < prefix > _ < noun > _ < verb > seems to be more common. So I personally would not change the word order.

> Question: it's not going to be possible for an i²c controller to serve 2+ i2c slave drivers, here it's one. Would that make sense to actually handle that use case?

Good point. I think we should do some research on this issue.

Reply:
Yes you are right, the controllers should be able to handle both master and slave at the same time.

Today, the API permits adding any number of slaves, it's a matter of the I2C controller driver to handle that, and it's proven in my i2c_virtual driver implementation.

The STM32 handle at least a maximum of 2 addresses, but I think there is some modes to ack a mask of adresses...

Reply:
Ok, but maybe you should decide with @tbursztyka  !

Reply:
In either case, would you want them to go in a separate ic2_slave.h header ?

Reply:
@tbursztyka what's your opinion on presence of `struct i2c_slave_driver_api` in this file after my clarifying comments?

@superna9999 I personally wouldn't create i2c_slave.h header file but keep everything here. That said I'm not an authoritative voice here.